src/beliefpropagation/sqrt_beliefpropagation.jl

```diff
@@ -0,0 +1,127 @@
+# using ITensors: scalartype
+# using ITensorNetworks: find_subgraph, map_diag, sqrt_diag, boundary_edges
+
+function sqrt_belief_propagation(
+  tn::ITensorNetwork,
+  mts::DataGraph;
+  niters=20,
+  # target_precision::Union{Float64,Nothing}=nothing,
+)
+  # compute_norm = target_precision == nothing ? false : true
+  sqrt_mts = sqrt_message_tensors(tn, mts)
+  for i in 1:niters
+    sqrt_mts, c = sqrt_belief_propagation_iteration(tn, sqrt_mts) #; compute_norm)
+    # if compute_norm && c <= target_precision
+    #   println(
+    #     "Belief Propagation finished. Reached a canonicalness of " *
+    #     string(c) *
+    #     " after $i iterations. ",
+    #   )
+    #   break
+    # end
+  end
+  return sqr_message_tensors(sqrt_mts)
+end
+
+function sqrt_belief_propagation_iteration(
+  tn::ITensorNetwork,
+  sqrt_mts::DataGraph;
+  # compute_norm=false,
+)
+  new_sqrt_mts = copy(sqrt_mts)
+  c = 0.0
+  es = edges(sqrt_mts)
+  for e in es
+    environment_tensornetworks = ITensorNetwork[
+      sqrt_mts[e_in] for
+      e_in in setdiff(boundary_edges(sqrt_mts, [src(e)]; dir=:in), [reverse(e)])
+    ]
+
+    new_sqrt_mts[src(e) => dst(e)] = update_sqrt_message_tensor(
+      tn, sqrt_mts[src(e)], environment_tensornetworks;
+    )
+
+    # if compute_norm
+    #   LHS, RHS = ITensors.contract(ITensor(sqrt_mts[src(e) => dst(e)])),
+    #   ITensors.contract(ITensor(new_sqrt_mts[src(e) => dst(e)]))
+    #   LHS /= sum(diag(LHS))
+    #   RHS /= sum(diag(RHS))
+    #   c += 0.5 * norm(LHS - RHS)
+    # end
```

Comment:
Does the same calculation of C not apply to sqrt_mts? I.e. we can't assess convergence criteria in the same way? If not then I guess for now we should just drop these lines of code and not have this function return `c`

Reply:
It's just a bit more subtle since the square root message tensors can have an arbitrary unitary gauge degree of freedom. We should be able to just square them of fix the gauge in some other way and then compare them from step to step, I just didn't implement that yet.

Reply:
That's fair. I suppose if we square them all up to test at every iteration that could be a fairly significant computational overhead so can see why you would leave it out.

Reply:
It's only `O(χ²)` so it should be sub-leading cost (except for linear networks) so I think it's ok, I just hadn't implemented it yet. I think it's good to put in, but I presume this will be an internal feature mostly for testing at first anyway since regular BP is faster.

Reply:
I'm going to leave this for a future PR.